### ISSUE TITLE
Insight for how I can learn better & translate pytorch code

### ISSUE THREAD

**Issue by danjavady** [LaurentMazare/tch-rs#545]
Hi,

I have several motivations for wanting to use Rust over Python&C++ when it comes to using Pytorch. However I am struggling to translate some code from python to rust here. I am new to both AI&Rust. 

For example, the code i am trying to translate: 
```nn.BatchNorm2d(32,1e-5,0.1,false)```

The Rust API is as follows:
```
pub fn batch_norm2d<'a, T: Borrow<super::Path<'a>>>(
    vs: T,
    out_dim: i64,
    config: BatchNormConfig,
) -> BatchNorm {
    batch_norm(vs, 2, out_dim, config)
}
```
is this the correct translation?
```
let config = BatchNormConfig { 
   cudnn_enabled: false, 
   eps: 1e-5,
   momentum: 0.1,
  ws_init: ??? 
  bs_init: ???
  
};
batch_norm2d(&path, 32, config)
```
What are the weights and biases for and what would I set them to in my context?
BatchNorm2d class in python 
```num_features, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True, device=None, dtype=None)```

How can I properly translate the above python code to rust? Config specifies the ep, momentum. But there is only one boolean in the Rust API where as there exists 2 in the python API. How can i properly map things out? 

Would you have any metaadvice on how I could learn how to translate Python code into Rust? I've been looking at the implementations but that has only taken me so far since I(seemingly?) can't see how everything is implemented in the native language of C++

**Comment by LaurentMazare** [comment]
The usual pattern is to use the `Default` trait for the config part and only override the values that you want set explicitely ([random example](https://github.com/LaurentMazare/tch-rs/blob/bb177d4d5f4d6d44fbe6b1f4206cbae93b282ad5/src/vision/inception.rs#L6)). The default values should match the Python implementation, this way you don't have to set `ws_init` or `bs_init` if you want the same setting as PyTorch.
When it comes to lining up parameters, they should be somewhat similar between the Python and Rust implementation so looking at the documentations to check the arguments names should help ([PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html), [docs.rs](https://docs.rs/tch/latest/tch/nn/fn.batch_norm2d.html), see also `BatchNormConfig`). In this case, the `affine` bit was not implemented on the Rust version (despite existing for `group_norm` and `layer_norm`) so I just pushed a small change adding support for it. Overall I would go with something like:
```rust
    let bn_config = nn::BatchNormConfig { eps: 1e-5, momentum: 0.1, affine: false, ..Default::default() };
    nn::batch_norm2d(&vs / "batch_norm", 32, bn_config)
```

**Comment by LaurentMazare** [comment]
Closing this now that there is a link to best practices in the faq section of the readme.
